Add example Rmarkdown notebook

jashapiro opened the pull request:
This PR adds an example notebook that takes some of the output from the counting script and generates some plots. 

The notebook ended up being pretty verbose and self-referential... I expect that we may end up teaching at least partly from the notebook itself, rather than just from slides.

I wasn't sure if there was any more content that might be good to add here, but I wanted to get the current state out for comment/review before going too much further (I probably went too far already).

One thing I don't cover is anything about params: I like them, but I feel like that is a slightly more advanced way of working with notebooks, or if not more advanced a bit of a different use case than the more interactive mode I wanted to focus on here.  Adding params means that the text itself has to be more generic, and that can cause some loss of context.
A thought I had was to add a paragraph at the end that discussed next steps or advanced topics that might include params and just general discussion of how the notebook might change... that might be something to do via discussion as well/instead.

I also moved the script & notebook to the root of the project, partly after some discussion with @jaclyn-taroni, and partly because I couldn't quite decide where the best place would be. Maybe we make a subfolder of `analysis` to put this workflow (Rscript, notebook & stitch script) in? Happy to hear thoughts. (I used `here:here()` for all the paths in the notebook to make moving it relatively painless).

I didn't include the html output as yet. I think we probably will want to include that in the repo, but in the interests of simplicity I did not include it in this PR as yet.

sjspielman commented on 02_plotting-mutation-counts.Rmd:
```diff
@@ -0,0 +1,265 @@
+
+We at the Data Lab really like using notebook documents like [Rmarkdown](https://rmarkdown.rstudio.com) (this document) and [Jupyter](https://jupyter.org) for development and data exploration.
+Notebooks allow mixing of text, code, and results in a single document that is easy to work with interactively.
+This allows much more freedom to include a lot of explanatory text, much like code comments, but with simpler syntax and the ability to include formatting and links more easily through the use of [markdown-formatted text](https://rmarkdown.rstudio.com/lesson-8.html) (or using the new visual editor in RStudio).
```
I'm a little/extremely pedantic about always using a noun after "this," but I'm not offended if you don't want to. 

```suggestion
Using notebooks allows much more freedom to include a lot of explanatory text, much like code comments, but with simpler syntax and the ability to include formatting and links more easily through the use of [markdown-formatted text](https://rmarkdown.rstudio.com/lesson-8.html) (or using the [new visual editor in RStudio](https://www.rstudio.com/blog/exploring-rstudio-visual-markdown-editor/)).
```

sjspielman commented on 02_plotting-mutation-counts.Rmd:
```diff
@@ -0,0 +1,265 @@
+
+**A note on formatting:**
+We like to keep our documents formatted with the philosophy of one sentence per line.
+This interacts well with git and GitHub, which record edits based on the line that changed.
```
```suggestion
This approach interacts well with git and GitHub, which record edits based on the line that changed.
```

sjspielman commented on 02_plotting-mutation-counts.Rmd:
```diff
@@ -0,0 +1,265 @@
+A nice solution we can use is the [`here` package](https://here.r-lib.org), which will allow us to specify paths relative to the project root directory.
+The project root is found by looking for an R project file with the `.Rproj` extension in a parent directory from the file (or a `.git` directory if there is no project file).
+
+We will also take advantage of the fact that the file.path() function (and `here::here()`), allows us to specify paths as a series of comma-separated partial paths, i.e., individual directories and files.
```
```suggestion
We will also take advantage of the fact that the file.path() function (and the `here::here()` function), allows us to specify paths as a series of comma-separated partial paths, i.e., individual directories and files.
```

jashapiro replied:
```suggestion
We will also take advantage of the fact that the `file.path()` function (and the `here::here()` function), allows us to specify paths as a series of comma-separated partial paths, i.e., individual directories and files.
```

sjspielman commented on 02_plotting-mutation-counts.Rmd:
```diff
@@ -0,0 +1,265 @@
+The project root is found by looking for an R project file with the `.Rproj` extension in a parent directory from the file (or a `.git` directory if there is no project file).
+
+We will also take advantage of the fact that the file.path() function (and `here::here()`), allows us to specify paths as a series of comma-separated partial paths, i.e., individual directories and files.
+If we knew everyone were using a UNIX machine, we could use a path like `analysis/gene_counts`, but this is a bit less portable to other systems.
```
```suggestion
If we knew everyone were using a UNIX machine, we could use a path like `analysis/gene_counts`, but this syntax is a bit less portable to other systems.
```

sjspielman commented on 02_plotting-mutation-counts.Rmd:
```diff
@@ -0,0 +1,265 @@
+
+We will also take advantage of the fact that the file.path() function (and `here::here()`), allows us to specify paths as a series of comma-separated partial paths, i.e., individual directories and files.
+If we knew everyone were using a UNIX machine, we could use a path like `analysis/gene_counts`, but this is a bit less portable to other systems.
+(Thankfully, paths like this *do* work on Windows now!) Using the separated list of elements also makes it easier to join partial paths together.
```
OSPL?

jashapiro replied:
```suggestion
(Thankfully, paths like this *do* work on Windows now!)
Using the separated list of elements also makes it easier to join partial paths together.
```

sjspielman commented on 02_plotting-mutation-counts.Rmd:
````diff
@@ -0,0 +1,265 @@
+
+```{r inputs}
+# Set the directory where the counts files are found
+counts_dir <- here::here("analysis", "gene_counts")
````
Oh I never actually realized it could be used like this! I have always done, `file.path(here::here(), "analysis", "gene_counts")`. So cool!

sjspielman commented on 02_plotting-mutation-counts.Rmd:
````diff
@@ -0,0 +1,265 @@
+}
+```
+
+### Define functions
````
Formatting in this section is a little funky with `:` and `-`?

sjspielman commented on 02_plotting-mutation-counts.Rmd:
```diff
@@ -0,0 +1,265 @@
+Finally, having the functions grouped together will make it easier if we decide to move them to an external file: - If the set of functions gets large, or if we want to use some functions across multiple notebooks, we can create a separate `.R` file (or files) and then use the `source()` command to load those functions into this notebook (or others) all at once.
+
+In this case, we will only have one function, which will make a plot of the number of samples with mutations for each gene.
+We'll give it some options to allow filtering to a minimum number of samples mutated, and highlighting of particular genes of interest.
```
```suggestion
We'll give it some options to allow filtering to a minimum number of mutated samples and highlighting particular genes of interest.
```

sjspielman commented on 02_plotting-mutation-counts.Rmd:
```diff
@@ -0,0 +1,265 @@
+
+Importantly, we will be sure to include some comments with documentation of the arguments for the function, as well as a description of what the function will return.
+
+To make this a bit more structured we will use a format called roxygen for these comments, which would be useful for automatically generating documentation if we were creating an R package.
```
```suggestion
To make these comments a bit more structured, we could use a format called *roxygen* for these comments, which would be useful for automatically generating documentation if we were creating an R package.
```
Or `roxygen`. Just some formatting to emphasize it.

sjspielman commented on 02_plotting-mutation-counts.Rmd:
```diff
@@ -0,0 +1,265 @@
+Importantly, we will be sure to include some comments with documentation of the arguments for the function, as well as a description of what the function will return.
+
+To make this a bit more structured we will use a format called roxygen for these comments, which would be useful for automatically generating documentation if we were creating an R package.
+We aren't doing that here, but it is still a nice enough convention, and RStudio has an option in the `Code` menu called `Insert Roxygen Skeleton` that can save us a bunch of typing and create the outline of what we need in a single step.
```
Something needs to be cleaned up for whether we use roxygen or not, because is it clearly used below?

jashapiro replied:
```suggestion
We aren't generating separate documentation files here, but it is still a nice enough convention, and RStudio has an option in the `Code` menu called `Insert Roxygen Skeleton` that can save us a bunch of typing and create the outline of what we need in a single step.
```

sjspielman commented on 02_plotting-mutation-counts.Rmd:
```diff
@@ -0,0 +1,265 @@
+    theme(axis.text.x = element_text(angle=45,hjust=1)) + 
+    scale_fill_brewer(palette = "Set1")
+    
+  return(plot_obj)
```
```suggestion
  # Return the plot
  return(plot_obj)
```

sjspielman commented on 02_plotting-mutation-counts.Rmd:
```diff
@@ -0,0 +1,265 @@
+
+**A note on code chunk names:** 
+You might have noticed that in most of the previous code chunks, I included some text in the `{r}` brackets on the first line of the chunk, which serves as the name of the chunk.
+The use of these names is a bit controversial among members of the Data Lab.
```
this is great! different ideas and philosophies are allowed!

sjspielman commented on 02_plotting-mutation-counts.Rmd:
```diff
@@ -0,0 +1,265 @@
+Some of us like them for the clarity and the fact that they allow easy navigation in RStudio.
+Others find that they get out of sync with the code inside them and can be more trouble than they are worth.
+In particular, no two chunks can have the same name, so sometimes duplicating or splitting a chunk can accidentally create a notebook that won't render properly.
+Your preferred method is up to you: here we are using names for chunks with defined purpose, but not for the more exploratory, iterative chunks in this section and the next.
```
I'd rearrange this - 
```suggestion
Here we are using names for chunks with defined purpose, but not for the more exploratory, iterative chunks in this section and the next, but your preferred method is up to you!
```

sjspielman reviewed (commented):
I really like how you have laid this out! I have some text suggestions throughout, but the overall content looks great.

One large-ish comment I have is `::` which is used in the R script but not here. I would keep these two files consistent on this front.

sjspielman reviewed (commented):
I really like how you have laid this out! I have some text suggestions throughout, but the overall content looks great.

One large-ish comment I have is `::` which is used in the R script but not here. I would keep these two files consistent on this front.

jashapiro commented on 02_plotting-mutation-counts.Rmd:
```diff
@@ -0,0 +1,265 @@
+Notebooks allow mixing of text, code, and results in a single document that is easy to work with interactively.
+This allows much more freedom to include a lot of explanatory text, much like code comments, but with simpler syntax and the ability to include formatting and links more easily through the use of [markdown-formatted text](https://rmarkdown.rstudio.com/lesson-8.html) (or using the new visual editor in RStudio).
+
+When you are ready to share, notebooks like this can easily be converted into a document such as `.html` (web page) or `.docx` (MS Word document) for sharing with others.
```
```suggestion
When you are ready to share your results with others, notebooks like this can easily be converted into a document such as `.html` (web page) or `.docx` (MS Word), results, plots, and all.
```

jashapiro commented on 02_plotting-mutation-counts.Rmd:
```diff
@@ -0,0 +1,265 @@
+### Saving plots
+
+If we are happy with the plots, we can save them.
+We didn't define the output plot file names at the top, just the directory, so we will create the file names now (using `file.path()` again!)
+.
```
```suggestion
We didn't define the output plot file names at the top, just the directory, so we will create the file names now (using `file.path()` again!).
```

sjspielman commented:
> Lots of thoughts here, and but no clear decision at this point, so I may ramble a bit...

:100: this me, rambling is welcome.

> I think I may have talked myself into moving them to project/analysis/mutation_counts/ (with no scripts subfolder). What do you think of that idea?

:tada: This is the one imo, except let's rename to `project/analysEs/mutation_counts/ `. Keeping the R code in the root directory gives the impression that the code is not organized at all but just dumped into a repo, and we want to emphasize structure. 

We can end up with this:

```
project/analyses/mutation_counts/01_count-gene-mutations.R
project/analyses/mutation_counts/02_plotting-mutation-counts.Rmd # and its HTML eventually
project/analyses/mutation_counts/run-mutation-analysis.R # eh?
```

To me, the `00`  (as in `project/analysis/mutation_counts/00_mutation-analysis.sh`) implies a pre-pre-processing step, so I tend to prefer `run` for this script similar to how PBTA is set up.




> I might add a comment to the project organization slides and/or readme about the project/scripts folder to note that this could be used for more "utility" scripts that might be part of multiple analyses?

👍

jashapiro commented:
> To me, the `00` (as in `project/analysis/mutation_counts/00_mutation-analysis.sh`) implies a pre-pre-processing step, so I tend to prefer `run` for this script similar to how PBTA is set up.
> 
> >

To me, `00` just means I want it to always sort to the top!  I don't want to have to search for the main run script.*  Maybe `00_run-mutation-analysis.sh` might work? (If there were pre-steps, I'd probably put them in a subdirectory, or maybe call them `00a`?)

*For workflow managers like snakemake, these tend to have standard names so not as much of an issue, but for shell scripts, less so.

jashapiro commented:
I think this is ready for another look. I added some commentary about packages (and changed to calling them packages, rather than libraries), moved things as discussed, and added a readme to the scripts directory.

sjspielman commented on analyses/mutation_counts/02_mutation-count-plots.Rmd:
````diff
@@ -0,0 +1,282 @@
+```{r}
+mb_plot <- plot_gene_mutations(mb_df, 
+                               highlight_genes = flags_df$gene, 
+                              highlight_title = "Frequently mutated")
````
spacing

sjspielman commented on analyses/mutation_counts/02_mutation-count-plots.Rmd:
````diff
@@ -0,0 +1,282 @@
+```{r packages}
+library(dplyr) # data wrangling tools
+library(ggplot2) # the plotting package we will use
+library(magrittr) # the "pipe" (%>%) we will use to facilitate sequential operations
````
Since `dplyr` imports this the `magrittr` load isn't strictly needed, but this approach also makes it more "verbose" that we're piping. I don't mind either way.

sjspielman commented on analyses/mutation_counts/02_mutation-count-plots.Rmd:
```diff
@@ -0,0 +1,282 @@
+During development, we might not know exactly what files are needed at the start, so this may not get *written* first, but having a consistent and prominent place where all paths are defined makes it easy to quickly determine what files are needed and where they will be found.
+
+Since this is an `Rmarkdown` file, all of the file paths will be relative to the location of this file.
+This means that if the file moves, the paths would move, which can be annoying!
```
```suggestion
This means that if the `Rmarkdown` file moves, the relative paths to files we are interacting with would change, which can be annoying and confusing!
```

sjspielman commented on analyses/mutation_counts/02_mutation-count-plots.Rmd:
```diff
@@ -0,0 +1,282 @@
+If we knew everyone were using a UNIX machine, we could use a path like `analysis/gene_counts`, but this syntax is a bit less portable to other systems.
+(Thankfully, paths like this *do* work on Windows now!)
+Using the separated list of elements also makes it easier to join partial paths together.
+We won't have to worry about whether a partial path ends with a separator slash or not; `file.path()` will add it as needed!
```
```suggestion
We won't have to worry about whether a partial path ends with a separator slash  (`/`) or not; `file.path()` will add it as needed!
```

jashapiro reviewed (commented):
Missed filing a few more comments here to look at when you rereview

sjspielman commented on analyses/mutation_counts/02_mutation-count-plots.Rmd:
````diff
@@ -0,0 +1,299 @@
+
+```{r outputs}
+# set the directory to save output plots
+plots_dir <- here::here("plots", "mutation_counts")
````
👍

sjspielman commented on analyses/mutation_counts/02_mutation-count-plots.Rmd:
```diff
@@ -0,0 +1,299 @@
+#' @param min_mutated The minimum `mutated_samples` value to include in the plot
+#'  (default: 3)
+#' @param highlight_genes A vector of genes to highlight in the plot (optional)
+#' @param highlight_title The title for the highlighted genes legend 
```
Indicate the default here

sjspielman reviewed (approved):
Looks about ready to me! I left a minor comment or two but don't need to see again.

I also see the rendered version has also been committed. I'm really agnostic on whether to include a knitted version or have them create it for the first time, but if you aren't agnostic, just make sure you're going the direction you prefer!

That said, if we _do_ include the rendered version here it will probably make for an exciting JS diff which could be useful to show them. Therefore I am now marginally in favor of keeping it in!

sjspielman commented:
I'm gonna note that I have 0 idea why github sometimes repeats review comments, and I just assume it's a github problem.

jashapiro commented:
> I also see the rendered version has also been committed. I'm really agnostic on whether to include a knitted version or have them create it for the first time, but if you aren't agnostic, just make sure you're going the direction you prefer!

My thought was they might want to open it to see the output when we started talking about it, possibly before rendering on their own. 🤷🏼
